File: java/stack/exercise_20.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> fila = new Stack<>();
        char[] chars = s.toCharArray();
        for(int i=0;i<chars.length;i++){
            if((chars[i] == '(') || (chars[i] == '{') || (chars[i] == '[')){
                fila.push(chars[i]);
            }else{
                if(fila.isEmpty()){
                    return false;
                }
                if(chars[i]==')'){
                    if(fila.pop()!='('){
                        return false;
                    }
                }else{
                    if(chars[i]=='}'){
                        if(fila.pop()!='{'){
                            return false;
                        }
                    }else{
                        if(chars[i]==']'){
                            if(fila.pop()!='['){
                                return false;
                            }
                        }
                    }
                }
            }
        }
        return fila.isEmpty();
    }
}

File: python/stack/exercise_20.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:
            return False
        stack = []
        for letter in s:
            if letter == "(" or letter == "{" or letter == "[":
                stack.append(letter)
            else:
                if not len(stack):
                    return False
                if (stack[-1] == "(" and letter == ")") or (stack[-1] == "{" and letter == "}") or (stack[-1] == "[" and letter == "]"):
                    stack.pop()
                else:
                    print([stack,letter])
                    return False
        if len(stack):
            return False
        return True
